### DISCUSSION TITLE
Allow Screen Space Reflections appear on Transparent Materials

### DISCUSSION THREAD

**Discussion by marcelb** [godotengine/godot-proposals#7274] [3D]
### Describe the project you are working on

A roque like game in an open ocean setting, where the player spends most of his time steering his ship across the waters.

### Describe the problem or limitation you are having in your project

My water material has severe limitations. It can either have SSReflections or be semi transparent. It can't have both.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

It should be possible to render SSReflections on semi transparent materials. 

Disclaimer: This is NOT about rendering semi transparent materials AS the reflection, but rendering the reflections of opaque materials ON a semi transparent material.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Water shaders like this should be possible:
![ApplicationFrameHost_fptJSg02Df](https://github.com/godotengine/godot-proposals/assets/308374/08ea5394-674a-46bc-a24b-f092749a25fb)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I tried to work it around by creating a custom shader that both does transparency and SSR, but its really hard. My main problem is, that I need the normal texture of the screen, which is not easily obtainable in Godot shaders, if at all.

### Is there a reason why this should be core and not an add-on in the asset library?

Great water shaders are a pillar of a lot of games, transport huge amounts of immersion and are commonly used. Both Unity and Unreals supports that, so Godot should join them.

**Comment by Calinou** [comment]
Moving to discussion, as there is no concrete technical implementation provided.

In the meantime, if you want reflections on transparent materials, use VoxelGI. This also supports off-screen reflections, which is often desired for reflections of large water bodies where screen-space artifacts can be noticeable.

SSR is pending a full rewrite at some point, as the current algorithm has lots of issues (both in terms of quality and performance). I'm not sure if the current implementation could be modified to be visible on transparent materials, but it may not be desired from a performance standpoint given how heavy the current implementation is.

A workaround you could use is to use dithered transparency for the water surface, but it won't look great at lower resolutions. This allows TAA and sharp SDFGI reflections to show up. This can be done using **Distance Fade > Pixel Dither** material property, or a custom shader. It could be made to look better with TAA enabled by jittering the dithering pattern every frame though. 

Another workaround is to have a ReflectionProbe parented to the camera, with its update mode set to **Always** and ambient mode set to **Disabled**. Set **Extents** to a high value like `(200, 200, 200)`, set **Max Distance** to `1000`, and you should get a nice-looking result that works with off-screen objects (though not always visually aligned with objects).

> My main problem is, that I need the normal texture of the screen, which is not easily obtainable in Godot shaders, if at all.

You can [get the normal roughness buffer](https://docs.godotengine.org/en/latest/tutorials/shaders/screen-reading_shaders.html#normal-roughness-texture) in Godot 4 when using the Forward+ rendering method. This is not available in Godot 3 though.

**Reply by marcelb** [reply]
Hey @Calinou, thanks for the huge amount of help and feedback.

VoxelGI sounds cool, even though I believe it might be more than I actually need? Nevertheless, I'll try.

SSR Rewrite sounds cool. It looks very low res rn, but at least for my use case that would be more than enough. I think there should be even a toggle materialwise whether the material receives SSR or not? That would be cool.

I didn't consider dithering. I usually don't like the look of it at all, but I'll give it a try.

I worked with ReflectionProbes a lot, it works extremly good for certain cases like rooms or reflective objects like spheres or boxes, but for large water surfaces I didn't find a way, that made sense.

Oh, the normal roughness buffer! I oversaw this one (I kinda only read "roughness" and thought its something else). I'll have a look, thats really cool.

**Comment by jams3223** [comment]
Is deferred rendering necessary in order to achieve this, particularly in addressing the various issues we have encountered with screen space reflection?
